define HB-JWT normatively (#262)
oauth-wg/oauth-selective-disclosure-jwt: draft-ietf-oauth-selective-disclosure-jwt.md

### Enabling Holder Binding {#enabling_holder_binding}

The Holder MAY add an optional JWT to prove Holder Binding to the Verifier.

#### Structure of the Holder Binding JWT {#hb-jwt}

This section defines the contents of the Holder Binding JWT.

The JWT MUST contain the following elements:
  * in the JOSE header,

>     * `typ`: REQUIRED. MUST be `hb+jwt`, which explicitly types the Holder Binding JWT as recommended in Section 3.11 of [@!RFC8725].

Reviewer: So... the encoded HB-JWT examples (or the code that generates them) should probably be updated to have this `typ`

Author: I updated the code and added examples, but did not re-generate example 4b, and 4a (was too much)....

    * `alg`: REQUIRED. A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. MUST NOT be `none` or an identifier for a symmetric algorithm (MAC).

>     * `kid`: `kid`: CONDITIONAL. JOSE Header containing the key ID. MUST NOT be present if `jwk` or `x5c` is present.

Reviewer: "`kid`: `kid`: "  should probably only have one "`kid`:"

>     * `jwk`: CONDITIONAL. JOSE Header containing the key material the new Credential shall be bound to. MUST NOT be present if `kid` or `x5c` is present.

Author: ```suggestion
    * `jwk`: CONDITIONAL. JOSE Header containing the key material. MUST NOT be present if `kid` or `x5c` is present.
```

>     * `jwk`: CONDITIONAL. JOSE Header containing the key material the new Credential shall be bound to. MUST NOT be present if `kid` or `x5c` is present.

Author: ok to remove JOSE headers and replace with a sentence saying "use the key identified in SD-JWT, if not clear which one, applications of sd-jwt must specify which key to use in HB-JWT."

>     * `x5c`: CONDITIONAL. JOSE Header containing a certificate or certificate chain corresponding to the key used to sign the JWT. This element MAY be used to convey a key attestation. In such a case, the actual key certificate will contain attributes related to the key properties. MUST NOT be present if `kid` or `jwk` is present.

Reviewer: I'm not sure I really grok the key attestation conveyance stuff, to be honest.

>     * `x5c`: CONDITIONAL. JOSE Header containing a certificate or certificate chain corresponding to the key used to sign the JWT. This element MAY be used to convey a key attestation. In such a case, the actual key certificate will contain attributes related to the key properties. MUST NOT be present if `kid` or `jwk` is present.

Reviewer: In general I think the JOSE headers that convey or point to the key don't need to be and shouldn't be in this list. The HB-JWT for an SD-JWT that is holder bound with a cnf.jwk doesn't need any of these, for one example. And the requirements re kid/jwk/x5c as written in these bullets now aren't totally clear (to me anyway) and need some fixup. Or just remove them.  I'd argue that the defined structure of the Holder Binding JWT should be only the minimal claims/headers necessary. Other requirements, if needed, such as saying that no more than one of `kid` or `jwk` or `x5c` be used together could be in text elsewhere.

  * in the JWT body,

>     * `iss`: REQUIRED. The issuer of the Holder Binding JWT. How the value is represented is up to the protocol used and out of scope of this specification.

Reviewer: I don't think issuer should be required. It's kinda implied by context to be the Holder. And doesn't really add anything to have it, that I can see. And there may not always be an obvious value to use for the iss value for the holder. Seems more likely to hurt than help.

Reviewer: Oh and also the current examples of the Holder Binding JWT or its payload have iat, aud, and nonce. But not iss. Removing it here, where it's not necessary IMHO, would be easier than updating the code and examples. 

Also I suspect there's some reason Dr. Fett didn't have `iss` in the HB JWT payloads. Even if not normative examples.

Reviewer 2: > Also I suspect there's some reason Dr. Fett didn't have `iss` in the HB JWT payloads. Even if not normative examples.

Essentially what you said above - (1) I'm not sure if there's always a defined value to put there and (2) I don't think it is necessary. Audience is really important, but I don't see a strict need for iss.

Author: ```suggestion
```

    * `iat`: REQUIRED. The value of this claim MUST be the time at which the Holder Binding JWT was issued using the syntax defined in [@!RFC7519].
    * `aud`: REQUIRED. The intended receiver of the Holder Binding JWT. How the value is represented is up to the protocol used and out of scope of this specification.
    * `nonce`: REQUIRED. Ensures the freshness of the signature. The value type of this claim MUST be a string. How this value is obtained is up to the protocol used and out of scope of this specification.

Below is a non-normative example of a Holder Binding JWT payload:

<{{examples/simple/hb_jwt_payload.json}}

> Below is a non-normative example of a JWT produced by signign a payload in the example above:

Reviewer 2: ```suggestion
Below is a non-normative example of a JWT produced by signing a payload in the example above:
```

The Holder MAY add an optional JWT to prove Holder Binding to the Verifier.

#### Structure of the Holder Binding JWT {#hb-jwt}

This section defines the contents of the Holder Binding JWT.

The JWT MUST contain the following elements:
  * in the JOSE header,
    * `typ`: REQUIRED. MUST be `hb+jwt`, which explicitly types the Holder Binding JWT as recommended in Section 3.11 of [@!RFC8725].
    * `alg`: REQUIRED. A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. MUST NOT be `none` or an identifier for a symmetric algorithm (MAC).
  * in the JWT body,
    * `iat`: REQUIRED. The value of this claim MUST be the time at which the Holder Binding JWT was issued using the syntax defined in [@!RFC7519].
    * `aud`: REQUIRED. The intended receiver of the Holder Binding JWT. How the value is represented is up to the protocol used and out of scope of this specification.
    * `nonce`: REQUIRED. Ensures the freshness of the signature. The value type of this claim MUST be a string. How this value is obtained is up to the protocol used and out of scope of this specification.

> To validate the signature on the Holder Binding JWT, the Verifier MUST use the key material in the SD-JWT. If it is not clear from the SD-JWT, HB-JWT MUST specify which key material the Verifier needs to use to validate HB-JWT using JOSE header parameters such as `kid` and `x5c`.

Reviewer 3: @Sakurann Does that mean I can use a DID as well? I don't want to be too restrictive in what can be used for verifying the HB-JWT? Being restrictive is the job of specific profiles such as the hi-assurance profile.

Author: I don't think anything in the text prevents DIDs. `kid` is expected to be used mainly with DIDs.

Reviewer 3: sounds good.

       2. Determine the public key for the Holder from the SD-JWT.
       3. Ensure that a signing algorithm was used that was deemed secure for the application. Refer to [@RFC8725], Sections 3.1 and 3.2 for details. The `none` algorithm MUST NOT be accepted.
       4. Validate the signature over the Holder Binding JWT.

>        5. Check that the Holder Binding JWT is valid using `iat`, and, if provided, `nbf` and `exp` claims.
>        6. Determine that the Holder Binding JWT is bound to the current transaction and was created for this Verifier (replay protection) by validating `nonce` and `aud` claims.

Author: ```suggestion
       5. Check that the Holder Binding JWT is valid using `iat`, and `typ` and, if provided, `nbf` and `exp` claims.
       6. Determine that the Holder Binding JWT is bound to the current transaction and was created for this Verifier (replay protection) by validating `nonce` and `aud` claims.
```

       2. Determine the public key for the Holder from the SD-JWT.
       3. Ensure that a signing algorithm was used that was deemed secure for the application. Refer to [@RFC8725], Sections 3.1 and 3.2 for details. The `none` algorithm MUST NOT be accepted.
       4. Validate the signature over the Holder Binding JWT.
       5. Check that the `typ` of the Holder Binding JWT is `hb+jwt`.
       6. Check that the creation time of the Holder Binding JWT, as determined by the `iat` claim, is within an acceptable window.

>        7. Check that the Holder Binding JWT is valid in all other respects, per [@!RFC7519] and [@!RFC8725].
>        7. Determine that the Holder Binding JWT is bound to the current transaction and was created for this Verifier (replay protection) by validating `nonce` and `aud` claims.

Reviewer: ```suggestion
       7. Determine that the Holder Binding JWT is bound to the current transaction and was created for this Verifier (replay protection) by validating `nonce` and `aud` claims.
       8. Check that the Holder Binding JWT is valid in all other respects, per [@!RFC7519] and [@!RFC8725].
```

       2. Determine the public key for the Holder from the SD-JWT.
       3. Ensure that a signing algorithm was used that was deemed secure for the application. Refer to [@RFC8725], Sections 3.1 and 3.2 for details. The `none` algorithm MUST NOT be accepted.
       4. Validate the signature over the Holder Binding JWT.

>        5. Check that the `typ` of the Holder Binding JWT is `hb+jwt`.

Reviewer 3: Why is `hb+jwt` needed for this? We would also need to register a media type -> application/hb+jwt for this.

Author: yes, that's the plan, to register

Author: I added the IANA registration section in this PR :)

>        6. Check that the Holder Binding JWT is valid using `iat`, and, if provided, `nbf` and `exp` claims.

Reviewer 3: If `nbf` and `exp` are not provided, then I think this step is not needed. How would somebody check whether the HB-JWT is valid if only `iat` is provided`

Reviewer: I think the idea is that the Verifier wants HB-JWTs that are relatively freshly created. And so would reject HB-JWTs that have an `iat` that falls outside some policy based window.

Author: it's the same logic as [DPoP JWT](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop-16#section-4.2). iat is sufficient to know if that HB-JWT is fresh enough.

Reviewer: I think the idea is that the Verifier wants HB-JWTs that are relatively freshly created. And so would reject HB-JWTs that have an `iat` that falls outside some policy based window.  But something like that should probably said more explicitly because `iat` doesn't have any normative validity rules on its own. Maybe, "Check that Holder Binding JWT was created within an acceptable timeframe as indicated by the `iat` claim" or something along those lines.

FWIW the idea is similar to DPoP and [Checking DPoP Proofs](https://www.ietf.org/archive/id/draft-ietf-oauth-dpop-16.html#name-checking-dpop-proofs) does have some text that attempts to say the the creation time of the JWT needs to be within an acceptable window.

And rather than mentioning the`nbf` and `exp` claims maybe have another general item like "Ensure that the Holder Binding JWT is valid in all other respects per [@!RFC7519]."

Reviewer 3: ok, got it

Author: ```suggestion
       6. Check that the creation time of the Holder Binding JWT, as determined by the `iat` claim, is within an acceptable window.
       7. Check that the Holder Binding JWT is valid in all other respects, per [@!RFC7519] and [@!RFC8725].
```
updated the verfication steps as suggested.
@awoie @bc-pi

Reviewer: wfm
